### ISSUE TITLE
LayerCake as a component

### ISSUE THREAD

**Issue by zzolo** [mhkeller/layercake#8]
So, I took a very basic crack at making LayerCake a component.
https://github.com/zzolo/layercake-ssr-testing/tree/e362b7ea8de4713f869b32b9aa550a37dc6bb761
Building and running it is the same as described in #5 .

The main LayerCake component and parts are in `layercake/`.  Note that I only implemented an SVG layer.  Also, I just sort of implemented this from scratch just as an example and was not going for parity with LayerCake as far as options and methods.
https://github.com/zzolo/layercake-ssr-testing/tree/e362b7ea8de4713f869b32b9aa550a37dc6bb761/layercake

Then, in `components/`, you can see it implemented in `Chart.html`, and this gets called with data in `Page.html`.
https://github.com/zzolo/layercake-ssr-testing/tree/e362b7ea8de4713f869b32b9aa550a37dc6bb761/components

The main difference is that we have to pass data around components, so there's many `{...props}` attributes in components, where `props` is just the full state of the current component.  For instance:

```
<SVG {...props} layers="{ svgLayers }" />
...
  computed: {
    props: state => state,
  }
```

This is not great, mostly because it's just kind of messy and it's not a two-way data source.  But, I think in practice it is not so bad.

There's also an open question in LayerCake's current form and here on how might be the best way to handle events like clicking.  I feel like the general idea is to create reusable components, such as an axis, and I am not sure of a good way to define a click handler as an option.  This is probably just a Svelte limitation where the data in a component doesn't have access to `this`, like a method does.

**Comment by mhkeller** [comment]
Ah cool — before I did the store I had been doing some `...props` stuff. The advantage to this component setup being you can create more cakes since you're not bound by having only one store, right? Does it help with any of the SSR stuff?

What kinds of events do you want to happen on click? For me, that would be where project-specific stuff takes over and you just put that into your component directly, i.e. it's risen above the level of what can be reliably reusable.

**Comment by zzolo** [comment]
I don't think the store limited the number of cakes, as each cake made its own store instance.  But maybe I am missing something.

This helps immensely with the SSR stuff such that there's no need for any logic around browser or no browser since Svelte can handle that in components naturally.

So, what I imagine, and an actual case I am looking at now, is clicking on a bar in a chart to then change some higher up property; for instance, using a chart as a sort of navigation.  Maybe not a common use case.

I imagine using LayerCake to make organization specific re-usable charts; for example a bar chart that when used could be controlled with some options.  So, that might look like:

```
<LineChart
  data="{ dataset }"
  chartOptions="{
    bar: {
      width: 10,
      onClick: (data, element) => {
        // Do something
      }
    }
  }"
>
```

Where `LineChart` would implement the `LayerCake` component.  Much like the example in the repo.

But maybe the way I am thinking about it is not really the intention, or I just need to start thinking in different ways.

**Comment by mhkeller** [comment]
@zzolo, I just rewrote this library to work with Svelte three and published it as version 2.0.0. The new API lays it out as a component and it's a huge improvement. I updated the website and docs. Check it out and let me know what you think.

**Comment by mhkeller** [comment]
I'll close this for now since it is now a component but lmk if you have any thoughts on this or a new issue!
